File: backend/controllers/workspace.py
import os
import platform
import subprocess
import shutil
from backend.storage import load_settings
from backend.controllers.git import all_repos

def open_in_editor(path):
    settings = load_settings()
    editor = settings.get('editor', 'code')
    if platform.system() == 'Windows':
        resolved_editor = shutil.which(editor) or editor
        subprocess.Popen([resolved_editor, path])
    elif platform.system() == 'Darwin' and editor in ('code', 'cursor', 'windsurf'):
        _DARWIN_APP = {'code': 'Visual Studio Code', 'cursor': 'Cursor', 'windsurf': 'Windsurf'}
        subprocess.Popen(['open', '-a', _DARWIN_APP[editor], path])
    else:
        subprocess.Popen([editor, path])

def handle_open(handler, params):
    target = params.get('path', [None])[0]
    if not target or not os.path.isdir(target):
        handler.send_json({'error': 'invalid path'}, 400)
        return
    open_in_editor(target)
    handler.send_json({'ok': True})

def handle_ls(handler, params):
    raw_path = params.get('path', ['~'])[0]

    # Windows: virtual drive listing
    if platform.system() == 'Windows' and raw_path == '__drives__':
        import string
        entries = []
        for letter in string.ascii_uppercase:
            drive = f"{letter}:\\"
            if os.path.exists(drive):
                entries.append({
                    'name': f"{letter}:",
                    'path': os.path.normpath(drive),
                    'is_git': False,
                    'in_workspace': False,
                })
        handler.send_json({'path': '__drives__', 'parent': None, 'entries': entries})
        return

    target = os.path.normpath(os.path.abspath(os.path.expanduser(raw_path)))
    if not os.path.isdir(target):
        handler.send_json({'error': 'not a directory'}, 400)
        return
    try:
        workspace_paths = {os.path.normcase(r['path']) for r in all_repos()}
        entries = []
        for e in sorted(os.scandir(target), key=lambda x: x.name):
            if not e.is_dir() or e.name.startswith('.'):
                continue
            norm_path = os.path.normpath(os.path.abspath(e.path))
            is_git = os.path.exists(os.path.join(norm_path, '.git'))
            entries.append({
                'name': e.name,
                'path': norm_path,
                'is_git': is_git,
                'in_workspace': os.path.normcase(norm_path) in workspace_paths,
            })
        parent_dir = os.path.dirname(target)
        parent = parent_dir if parent_dir != target else None
        # Windows: at drive root, allow navigating up to drive list
        if parent is None and platform.system() == 'Windows':
            parent = '__drives__'
        handler.send_json({'path': target, 'parent': parent, 'entries': entries})
    except PermissionError:
        handler.send_json({'error': 'permission denied'}, 403)

File: internal/controllers/workspace/workspace.go
// Package workspace implements the directory browser (/api/ls) and editor open
// (/api/open) endpoints.
package workspace

import (
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"

	gitctl "github.com/imohiyoko/devhub/internal/controllers/git"
	"github.com/imohiyoko/devhub/internal/httpx"
	"github.com/imohiyoko/devhub/internal/pathutil"
	"github.com/imohiyoko/devhub/internal/platform"
)

// darwinApps maps an editor id to its macOS application name for `open -a`.
var darwinApps = map[string]string{"code": "Visual Studio Code", "cursor": "Cursor", "windsurf": "Windsurf"}

// settingsReader is the narrow persistence the workspace controller needs: it
// only reads the shared settings document (for the configured editor). It reads
// the global document rather than owning a keyspace, so it depends on the typed
// LoadSettings helper, not the raw key/value seam. *storage.Store satisfies it.
type settingsReader interface {
	LoadSettings() (map[string]any, error)
}

// Controller serves workspace endpoints; it consults git for the repo list.
type Controller struct {
	store settingsReader
	git   *gitctl.Controller
}

// New returns a workspace controller.
func New(store settingsReader, git *gitctl.Controller) *Controller {
	return &Controller{store: store, git: git}
}

func (c *Controller) editor() string {
	settings, _ := c.store.LoadSettings()
	if e, ok := settings["editor"].(string); ok && e != "" {
		return e
	}
	return "code"
}

// OpenInEditor opens path in the configured editor (exported for the env-launcher).
func (c *Controller) OpenInEditor(path string) { openInEditor(c.editor(), path) }

func openInEditor(editor, path string) {
	var cmd *exec.Cmd
	switch runtime.GOOS {
	case "windows":
		resolved := editor
		if p, err := exec.LookPath(editor); err == nil {
			resolved = p
		}
		cmd = exec.Command(resolved, path) //execaudit:workspace-editor
	case "darwin":
		if app, ok := darwinApps[editor]; ok {
			cmd = exec.Command("open", "-a", app, path) //execaudit:workspace-editor
		} else {
			cmd = exec.Command(editor, path) //execaudit:workspace-editor
		}
	default:
		cmd = exec.Command(editor, path) //execaudit:workspace-editor
	}
	_ = cmd.Start() // fire-and-forget, matching subprocess.Popen
}

// HandleOpen serves GET /api/open?path=<dir>: opens the directory in the editor.
// The path is used as-is (the frontend sends absolute paths from /api/ls).
func (c *Controller) HandleOpen(w http.ResponseWriter, r *http.Request) error {
	target := r.URL.Query().Get("path")
	if target == "" || !pathutil.IsDir(target) {
		return httpx.Errorf(http.StatusBadRequest, "invalid path")
	}
	openInEditor(c.editor(), target)
	httpx.WriteJSON(w, http.StatusOK, map[string]any{"ok": true})
	return nil
}

// HandleLs serves GET /api/ls?path=<dir>: lists subdirectories with git/workspace
// annotations. Defaults to ~; on Windows the sentinel __drives__ lists drives.
func (c *Controller) HandleLs(w http.ResponseWriter, r *http.Request) error {
	raw := r.URL.Query().Get("path")
	if raw == "" {
		raw = "~"
	}

	if platform.IsWindows() && raw == "__drives__" {
		httpx.WriteJSON(w, http.StatusOK, map[string]any{
			"path": "__drives__", "parent": nil, "is_git": false, "entries": listDrives(),
		})
		return nil
	}

	target := pathutil.AbsExpand(raw)
	if !pathutil.IsDir(target) {
		return httpx.Errorf(http.StatusBadRequest, "not a directory")
	}

	wsPaths := map[string]bool{}
	for _, rp := range c.git.AllRepos() {
		wsPaths[pathutil.NormCase(rp.Path)] = true
	}

	dirents, err := os.ReadDir(target) // sorted by name
	if err != nil {
		if os.IsPermission(err) {
			return httpx.Errorf(http.StatusForbidden, "permission denied")
		}
		return err
	}
	entries := []map[string]any{}
	for _, e := range dirents {
		name := e.Name()
		if strings.HasPrefix(name, ".") {
			continue
		}
		full := filepath.Join(target, name)
		if !pathutil.IsDir(full) { // follows symlinks, like scandir.is_dir()
			continue
		}
		entries = append(entries, map[string]any{
			"name":         name,
			"path":         full,
			"is_git":       pathutil.Exists(filepath.Join(full, ".git")),
			"in_workspace": wsPaths[pathutil.NormCase(full)],
		})
	}

	parentDir := filepath.Dir(target)
	var parent any
	if parentDir != target {
		parent = parentDir
	} else if platform.IsWindows() {
		parent = "__drives__"
	}

	httpx.WriteJSON(w, http.StatusOK, map[string]any{
		"path":    target,
		"parent":  parent,
		"is_git":  pathutil.Exists(filepath.Join(target, ".git")),
		"entries": entries,
	})
	return nil
}

// listDrives probes A:..Z: (avoids a Windows-only syscall dependency).
func listDrives() []map[string]any {
	entries := []map[string]any{}
	for c := 'A'; c <= 'Z'; c++ {
		root := string(c) + `:\`
		if pathutil.Exists(root) {
			entries = append(entries, map[string]any{
				"name":         string(c) + ":",
				"path":         filepath.Clean(root),
				"is_git":       false,
				"in_workspace": false,
			})
		}
	}
	return entries
}
